mac/Keyman4MacIM/Keyman4MacIM/KMInputMethodEventHandler.m

```diff
@@ -723,98 +452,102 @@ -(NSString*)readContext:(NSEvent *)event forClient:(id) client {
+  
+  if (output.alert) {
+    NSBeep();
   }
-  switch(action.actionType) {
-    case AlertAction:
-      NSBeep();
-      break;
-    case PersistOptionAction:
-      if(action.key && action.value) {
-        [self.appDelegate logDebugMessage:@"PersistOptionAction calling writePersistedOptions, key: %@, value: %@", action.key, action.value];
-        [self.appDelegate writePersistedOptions:action.key withValue:action.value];
+}
+
+-(void) persistOptions:(NSDictionary*)options{
+  if(options.count>0) {
+    for(NSString *key in options) {
+      NSString *value = [options objectForKey:key];
+      if(key && value) {
+        [self.appDelegate logDebugMessage:@"applyNonTextualOutput calling writePersistedOptions, key: %@, value: %@", key, value];
+        [self.appDelegate writePersistedOptions:key withValue:value];
       }
       else {
-        [self.appDelegate logDebugMessage:@"Invalid values for PersistOptionAction, not writing to UserDefaults, key: %@, value: %@", action.key, action.value];
+        [self.appDelegate logDebugMessage:@"applyNonTextualOutput, invalid values in optionsToPersist, not writing to UserDefaults, key: %@, value: %@", key, value];
       }
-      break;
-    case InvalidateContextAction:
-      self.contextChanged = YES;
-      break;
-    case CapsLockAction: {
-      //TODO: handle this somewhere
-      break;
     }
-    default:
-      break;
   }
 }
 
--(void)executeCompositeOperation:(KMActionOperation*)operation keyDownEvent:(nonnull NSEvent *)event client:(id) client {
-  [self.appDelegate logDebugMessage:@"KXMInputMethodHandler executeCompositeOperation, composite operation: %@", operation];
-  
-  if (operation.isTextOnlyScenario) {
-    [self.appDelegate logDebugMessage:@"KXMInputMethodHandler executeCompositeOperation, text only scenario"];
-    [self insertAndReplaceTextForOperation:operation client:client];
-  } else if (operation.isBackspaceOnlyScenario) {
-    [self.appDelegate logDebugMessage:@"KXMInputMethodHandler executeCompositeOperation, backspace only scenario"];
-    [self sendEventsForOperation: event actionOperation:operation];
-  } else if (operation.isTextAndBackspaceScenario) {
-    if (self.apiCompliance.mustBackspaceUsingEvents) {
-      [self.appDelegate logDebugMessage:@"KXMInputMethodHandler executeCompositeOperation, text and backspace scenario with events"];
-     [self sendEventsForOperation: event actionOperation:operation];
-    } else {
-      [self.appDelegate logDebugMessage:@"KXMInputMethodHandler executeCompositeOperation, text and backspace scenario with insert API"];
-      // directly insert text and handle backspaces by using replace
-      [self insertAndReplaceTextForOperation:operation client:client];
-    }
+-(void) applyCapsLock:(CapsLockState)capsLockState {
+  switch(capsLockState) {
+    case On:
+      //TODO: handle this
+      break;
+    case Off:
+      //TODO: handle this
+      break;
+    case Unchanged:
+      // do nothing
+      break;
```

Comment:
Do we have an issue to track this? If so, can you put it into the TODO comments?

Reply:
created #10244